skip mechanism for depending tests

MichalKlacansky opened the pull request:
add possibility when test fails depending tests will be skipped

mentalisttraceur commented:
Thank you for the contribution! I think I like the overall idea, and I really appreciate that you took the time to match the docs and everything.

I gotta think over if this is the best interface for it.

At a minimum, three nits:

1. This is a new feature, so per SemVer the new version would be 3.1.0.
2. "Depends on test skip" is too not-self-explanatory as a chain of words, so I gotta think more about the name.
3. If we do it this way, then the skip message should indicate the different skip reason: "was skipped because dependency failed" rather than just "was skipped".

I'm gonna try to give this a full think (and load up all my thoughts on #6 because that seems relevant) sometime this week and follow up with at least something sometime this or next week.

MichalKlacansky commented:
good points :)
Fixed version
changed "Depends on test skip" to "Skip depending test on fail"
changed skip reason to: "was skipped because dependency failed."
and little refactor so this message can be displayed

mentalisttraceur commented:
Sorry for taking two weeks longer than I expected. Finally have some time to do that thinkover I was intending, so that's starting now.

I also just ran into a use-case for exactly this functionality while doing some work on another thing, so I'm a lot more vividly excited+motivated for it now.

mentalisttraceur commented:
Question: do you *ever* want dependencies to cascade failures, or would you be having all of them do the skip-if-dependency-failed behavior?

mentalisttraceur commented on DependencyLibrary.py:
```diff
@@ -22,6 +22,8 @@ def _depends_on(status_map, dependency_type, name):
     if status == 'SKIP':
         raise _SkipExecution(message + ' was skipped.')
+    if skip:
+        raise _SkipExecution(message + ' was skipped because dependency failed.')
```
Now that I'm actually looking at this closely in a brain-on way, I'm realizing that the correct skip message ending is just `' failed.'`

We already know *this test* is being skipped because the status is skip, the message is telling us *why* this status is happening: in this case, this test is skipping because the dependency failed.

mentalisttraceur commented on README.rst:
```diff
@@ -135,3 +135,24 @@ If you make a test depend on itself or on the suite that contains it,
+
+The skip message follows this format::
+
+    Dependency not met: test case 'A Test that Depends on "Failing Test"' was skipped because dependency failed.
```
Also over here: the actual output will be:

        Dependency not met: test case 'Failing Test' [...]

The current logic would not produce

    Dependency not met: test case 'A Test that Depends on "Failing Test"' [...]

because the `name` variable in that message is the name of the dependency (the argument to `Depends on test [...]`).

mentalisttraceur commented:
In the meantime, since I'm a pretty *slow* maintainer due to how I like to think things through and my limited time, you might find this useful in the interim: 

```
*** Keywords ***
Depends on test
    [Arguments]  ${test_name}
    ${test_name_repr} =  Evaluate  repr($test_name)
    TRY
        DependencyLibrary.depends on test  ${test_name}
    EXCEPT  Dependency not met: test case ${test_name_repr} failed.
        Skip    Dependency not met: test case ${test_name_repr} failed.
```

I didn't test but I think that would let you polyfill the behavior you want in your test suites while I putz around with Design Thoughts(tm).

mentalisttraceur commented:
So to the best of my recollection, the design originally went like this:

1. At first, Robot Framework didn't even *have* a "skip" ability. There was only pass or fail. So I made dependencies fail, because fail-fast is more useful than try-to-run-and-do-things-that-will-fail.

2. Then "skip" status came along and I remember deciding that skip->skip instead of skip->fail was the obviously right approach ... but I don't remember what my thought process was about fail->fail vs fail->skip.

The only good argument for failure->failure I can think of *right now* is:

1. It more accurately counts the fraction of our features that's broken due to dependencies.

I can also think of a really tenuous and kinda mean-spirited argument:

2. The experience creates some incentive to have less dependencies between tests. (However, this argument is only true along the *statefulness* dimension of dependency... there is also the *semantic* dimension, which is arguably more important, where it's less about the bad design of "I need that test to mutate shared state so that this test can work" and more about "this functionality builds on the functionality tested in that test - there's no point in testing this if that fails"... and along the semantic dimension, I think fail->skip is by far more sensible.)

So, on the side of fail->skip:

1. When we're looking at test failures, signal-to-noise is very important. A test which fails just because something else failed is almost 100% noise. It really really sucks to scroll past verbose logs from errors that just tell us that an earlier step broke when we want to know the error from that earlier step.

2. For the most part, everything interesting about dependencies is already known statically in the source (we don't get much value from seeing 10 tests fail instead of 1 fail and 9 more than usual get skipped, and we can count skip statuses that match "Dependency not met").

mentalisttraceur commented:
Yeah I'm really leaning towards making all of the `Depends on ...` keywords automatically skip rather than fail on failed dependencies.

Especially since it clearly doesn't conflict with #6, and can *probably* be made to work with #5 (that one is trickier because I was imagining a simple retry scheme on the `pabot` side... I gotta start reviewing my thoughts about that).

mentalisttraceur commented:
One thing that this brings up, is the distinction between depending on a test, and depending on functionality proven in the test.

Luckily, I think I have already convinced myself that for *both* of those two, it is good to have failure of a dependency imply skip of the dependent.

mentalisttraceur commented:
Oh you know what? There is a big chance that I did fail->fail instead of fail->skip for backwards compatibility.

Because that way I could have all uses of `robot.api.SkipExecution` tucked away into code paths that would only run on Robot Framework versions new enough to have skip status.

It's hard to overstate just how severely bent towards portability and backwards-compatibility my mind was (and still kinda is)!

I'm feeling more okay with totally shedding that - just dropping support for Robot Framework versions without a `SkipExecution` exception.

mentalisttraceur commented:
Fixed in v4.0.0 - now a failed dependency always causes skip instead of failure.

mentalisttraceur commented:
One interesting caveat is that Robot code can easily turn a keyword failure into a test skip, but it's a lot harder to catch a skip and turn it into a fail.

This only matters if people are using `Depends on ...` keywords as building blocks for more complex keywords... which I certainly might've recommended in some cases.

But I think if there's enough use-cases and demand for that still, we can just add another keyword like `Check dependency`.
